`INSERT` into `ARRAY(OBJECT(IGNORED))` fails if data types are inconsistent

### CrateDB version

6.2.1, nightly-6.3.0-2026-02-25-00-02

### CrateDB setup information

Standard single node

### Problem description

Inserting into an `ARRAY(OBJECT(IGNORED))` fails if data types are inconsistent. It used to work on CrateDB 5.10.16.

### Steps to Reproduce

```sql
CREATE TABLE IF NOT EXISTS doc.t1 (
   "document" OBJECT(DYNAMIC) AS (
      "value" OBJECT(DYNAMIC) AS (
         "readings" ARRAY(OBJECT(IGNORED))
      )
   )
);

INSERT INTO doc.t1 VALUES (
$$
{
    "value": {
        "readings": [
            {
                "locations": [
                    {
                        "attr_1": "text"
                    },
                    {
                        "attr_1": ["text", "array"]
                   }
                ]
            }
        ]
    }
}
$$);
```

### Actual Result

```
io.crate.exceptions.SQLParseException: 'text' is not convertible to 'text_array'
	at io.crate.exceptions.SQLExceptions.esToCrateException(SQLExceptions.java:200)
	at io.crate.exceptions.SQLExceptions.prepareForClientTransmission(SQLExceptions.java:189)
	at io.crate.rest.action.SqlHttpHandler.sendResponse(SqlHttpHandler.java:175)
	at io.crate.rest.action.SqlHttpHandler.lambda$channelRead0$0(SqlHttpHandler.java:126)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:884)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:862)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2221)
	at io.crate.rest.action.RestRowCountReceiver.fail(RestRowCountReceiver.java:95)
	at io.crate.session.RowConsumerToResultReceiver.accept(RowConsumerToResultReceiver.java:67)
	at io.crate.planner.operators.InsertFromValues.lambda$executeOrFail$2(InsertFromValues.java:301)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:884)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:862)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531)
	at java.base/java.util.concurrent.CompletableFuture.completeExceptionally(CompletableFuture.java:2221)
	at io.crate.planner.operators.InsertFromValues.lambda$execute$0(InsertFromValues.java:657)
	at io.crate.planner.operators.InsertFromValues$1.onResponse(InsertFromValues.java:695)
	at io.crate.planner.operators.InsertFromValues$1.onResponse(InsertFromValues.java:682)
	at org.elasticsearch.action.support.RetryableAction$RetryingListener.onResponse(RetryableAction.java:139)
	at org.elasticsearch.action.ActionListener.accept(ActionListener.java:53)
	at org.elasticsearch.action.ActionListener.accept(ActionListener.java:38)
	at java.base/java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:884)
	at java.base/java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:862)
	at java.base/java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:531)
	at java.base/java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2206)
	at io.crate.concurrent.FutureActionListener.onResponse(FutureActionListener.java:32)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase.finishOnSuccess(TransportReplicationAction.java:820)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:743)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$ReroutePhase$1.handleResponse(TransportReplicationAction.java:729)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.processResponse(TransportService.java:924)
	at org.elasticsearch.transport.TransportService$DirectResponseChannel.sendResponse(TransportService.java:904)
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:40)
	at org.elasticsearch.action.support.ChannelActionListener.onResponse(ChannelActionListener.java:28)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.lambda$runWithPrimaryShardReference$0(TransportReplicationAction.java:386)
	at org.elasticsearch.action.ActionListener$SimpleListener.onResponse(ActionListener.java:96)
	at org.elasticsearch.action.ActionListener.lambda$map$0(ActionListener.java:64)
	at org.elasticsearch.action.ActionListener.lambda$withOnResponse$0(ActionListener.java:114)
	at org.elasticsearch.action.ActionListener$SimpleListener.onResponse(ActionListener.java:96)
	at org.elasticsearch.action.support.replication.ReplicationOperation.finish(ReplicationOperation.java:350)
	at org.elasticsearch.action.support.replication.ReplicationOperation.decPendingAndFinishIfNeeded(ReplicationOperation.java:331)
	at org.elasticsearch.action.support.replication.ReplicationOperation$1.onResponse(ReplicationOperation.java:159)
	at org.elasticsearch.action.support.replication.ReplicationOperation$1.onResponse(ReplicationOperation.java:151)
	at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.maybeFinish(TransportWriteAction.java:200)
	at org.elasticsearch.action.support.replication.TransportWriteAction$AsyncAfterWriteAction.run(TransportWriteAction.java:216)
	at org.elasticsearch.action.support.replication.TransportWriteAction$WritePrimaryResult.runPostReplicationActions(TransportWriteAction.java:137)
	at org.elasticsearch.action.support.replication.ReplicationOperation.handlePrimaryResult(ReplicationOperation.java:151)
	at org.elasticsearch.action.ActionListener$SimpleListener.onResponse(ActionListener.java:96)
	at io.crate.execution.dml.TransportShardAction.shardOperationOnPrimary(TransportShardAction.java:118)
	at io.crate.execution.dml.TransportShardAction.shardOperationOnPrimary(TransportShardAction.java:56)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$PrimaryShardReference.perform(TransportReplicationAction.java:890)
	at org.elasticsearch.action.support.replication.ReplicationOperation.execute(ReplicationOperation.java:123)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.runWithPrimaryShardReference(TransportReplicationAction.java:401)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.lambda$doRun$0(TransportReplicationAction.java:323)
	at org.elasticsearch.action.ActionListener$SimpleListener.onResponse(ActionListener.java:96)
	at org.elasticsearch.index.shard.IndexShard.lambda$wrapPrimaryOperationPermitListener$0(IndexShard.java:2622)
	at org.elasticsearch.action.ActionListener.lambda$withOnResponse$0(ActionListener.java:114)
	at org.elasticsearch.action.ActionListener$SimpleListener.onResponse(ActionListener.java:96)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:293)
	at org.elasticsearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:246)
	at org.elasticsearch.index.shard.IndexShard.acquirePrimaryOperationPermit(IndexShard.java:2598)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.acquirePrimaryOperationPermit(TransportReplicationAction.java:838)
	at org.elasticsearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:319)
	at org.elasticsearch.common.util.concurrent.RejectableRunnable.run(RejectableRunnable.java:36)
	at org.elasticsearch.action.support.replication.TransportReplicationAction.handlePrimaryRequest(TransportReplicationAction.java:285)
	at org.elasticsearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:59)
	at org.elasticsearch.transport.TransportService$2$1.doRun(TransportService.java:119)
	at org.elasticsearch.common.util.concurrent.RejectableRunnable.run(RejectableRunnable.java:36)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1090)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:614)
	at java.base/java.lang.Thread.run(Thread.java:1474)
Caused by: java.lang.IllegalArgumentException: 'text' is not convertible to 'text_array'
	at io.crate.types.DataType.merge(DataType.java:264)
	at io.crate.types.ArrayType.merge(ArrayType.java:392)
	at io.crate.types.DataTypes.merge(DataTypes.java:633)
	at io.crate.types.DataTypes.merge(DataTypes.java:620)
	at java.base/java.util.HashMap.merge(HashMap.java:1392)
	at io.crate.types.ObjectType$Builder.mergeInnerType(ObjectType.java:136)
	at io.crate.types.ObjectType.merge(ObjectType.java:354)
	at io.crate.types.ObjectType.merge(ObjectType.java:339)
	at io.crate.types.DataTypes.merge(DataTypes.java:633)
	at io.crate.types.DataTypes.merge(DataTypes.java:620)
	at io.crate.types.DataTypes.typeFromList(DataTypes.java:376)
	at io.crate.types.DataTypes.guessType(DataTypes.java:316)
	at io.crate.types.DataTypes.guessType(DataTypes.java:312)
	at io.crate.types.DataTypes.typeFromList(DataTypes.java:375)
	at io.crate.types.DataTypes.guessType(DataTypes.java:316)
	at io.crate.execution.dml.DynamicIndexer.guessType(DynamicIndexer.java:161)
	at io.crate.execution.dml.ObjectIndexer.addNewColumns(ObjectIndexer.java:268)
	at io.crate.execution.dml.ObjectIndexer.collectSchemaUpdates(ObjectIndexer.java:209)
	at io.crate.execution.dml.ObjectIndexer.collectSchemaUpdates(ObjectIndexer.java:55)
	at io.crate.execution.dml.ObjectIndexer.collectSchemaUpdates(ObjectIndexer.java:201)
	at io.crate.execution.dml.ObjectIndexer.collectSchemaUpdates(ObjectIndexer.java:55)
	at io.crate.execution.dml.Indexer.collectSchemaUpdates(Indexer.java:904)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.insert(TransportShardUpsertAction.java:541)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.insertOrUpdate(TransportShardUpsertAction.java:398)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.processRequestItems(TransportShardUpsertAction.java:198)
	at io.crate.execution.dml.upsert.TransportShardUpsertAction.processRequestItems(TransportShardUpsertAction.java:89)
	at io.crate.execution.dml.TransportShardAction$1.call(TransportShardAction.java:115)
	at io.crate.execution.dml.TransportShardAction$1.call(TransportShardAction.java:111)
	at io.crate.execution.dml.TransportShardAction.withActiveOperation(TransportShardAction.java:146)
	... 23 more
```

### Expected Result

Successful execution (as with CrateDB 5.10.16).

--- a/docs/appendices/release-notes/6.2.2.rst
+++ b/docs/appendices/release-notes/6.2.2.rst
@@ -104,3 +104,14 @@ Fixes
       SELECT unnest([ {child_obj = { known_col = 'foo'}} ]) AS obj_arr
     ) AS sub1;
 
+- Fixed a regression introduced in :ref:`version_6.0.0` that caused an error
+  when trying to insert a nested array, or JSON string containing a nested
+  array, with incompatible inner value types into a column of type
+  ``OBJECT(ARRAY(OBJECT(IGNORED)))``. Example::
+
+    CREATE TABLE t1 (obj OBJECT(DYNAMIC) AS (arr ARRAY(OBJECT(IGNORED))));
+    INSERT INTO t1 (obj) VALUES (
+    $$
+       {"arr":[{"b": "str_val"}, {"b": ["array_val"]}]}
+    $$);
+

--- a/server/src/main/java/io/crate/execution/dml/ObjectIndexer.java
+++ b/server/src/main/java/io/crate/execution/dml/ObjectIndexer.java
@@ -265,14 +265,24 @@ public class ObjectIndexer implements ValueIndexer<Map<String, Object>> {
         for (var entry : value.entrySet()) {
             String innerName = entry.getKey();
             Object innerValue = entry.getValue();
-            DataType<?> type = DynamicIndexer.guessType(innerValue);
             ObjectIndexer.Child child = children.get(innerName);
             if (child != null) {
                 var childType = child.reference.valueType();
                 // We allow overwriting undefined typed children with new typed ones
-                if (!(childType instanceof UndefinedType) || type instanceof UndefinedType) {
+                if (!(childType instanceof UndefinedType)) {
                     continue;
                 }
+            }
+            // Type guessing is only required if no child exists or is undefined.
+            // On a nested list of map values with different map element types, guessing would raise an exception while
+            // trying to guess a common array element type (map with same inner types) even that the related map/object
+            // column policy is IGNORED.
+            DataType<?> type = DynamicIndexer.guessType(innerValue);
+            if (type instanceof UndefinedType && child != null) {
+                assert child.reference.valueType() instanceof UndefinedType
+                    : "Should only be reached if the existing child is undefined";
+                // Child exists already with undefined type
+                continue;
             }
             if (isStrict) {
                 throw new IllegalArgumentException(String.format(

--- a/server/src/test/java/io/crate/execution/dml/IndexerTest.java
+++ b/server/src/test/java/io/crate/execution/dml/IndexerTest.java
@@ -30,7 +30,6 @@ import static io.crate.types.GeoShapeType.Names.TREE_BKD;
 import static io.crate.types.GeoShapeType.Names.TREE_GEOHASH;
 import static io.crate.types.GeoShapeType.Names.TREE_LEGACY_QUADTREE;
 import static io.crate.types.GeoShapeType.Names.TREE_QUADTREE;
-import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 import static org.elasticsearch.cluster.metadata.Metadata.OID_UNASSIGNED;
 
@@ -1833,6 +1832,34 @@ public class IndexerTest extends CrateDummyClusterServiceUnitTest {
         ParsedDocument parsedDoc = indexer.index(item1, updatedTable.rootColumns());
         assertThat(source(parsedDoc, updatedTable)).isEqualTo("{\"points\":[{\"x\":1},{\"x\":null}]}");
         assertTranslogParses(parsedDoc, updatedTable);
+    }
+
+    @Test
+    public void test_can_index_nested_array_of_ignored_objects_with_mixed_inner_types() throws Exception {
+        SQLExecutor e = SQLExecutor.of(clusterService)
+            .addTable("CREATE TABLE tbl (o OBJECT AS (arr ARRAY(OBJECT (IGNORED))))");
+        DocTableInfo table = e.resolveTableInfo("tbl");
+
+        var indexer = getIndexer(e, "tbl", "o");
+        IndexItem item = item(Map.of("arr", List.of(Map.of("k1", List.of(Map.of("k2", "text"), Map.of("k2", List.of("text", "array")))))));
+
+        // check that schema updates do not cause errors, nothing should be collected since the inner object elements are ignored
+        List<Reference> newColumns = indexer.collectSchemaUpdates(item);
+        assertThat(newColumns).isEmpty();
+
+        // Ensure that the document is indexed and parsed correctly, even if the inner object contains mixed types (string and array)
+        ParsedDocument doc = indexer.index(item);
+        assertThat(source(doc, table)).isEqualToIgnoringWhitespace(
+            """
+            {"o":{"arr":[{"k1":[{"k2":"text"},{"k2":["text","array"]}]}]}}
+            """
+        );
+        assertThat(doc.doc().getFields("o.arr")).isEmpty();
+        assertThat(doc.doc().getFields("o.arr.k1")).isEmpty();
+        assertThat(doc.doc().getFields())
+            .as("source, seqNo, id...")
+            .hasSize(10);
+        assertTranslogParses(doc, table);
     }
 
     public static void assertTranslogParses(ParsedDocument doc, DocTableInfo info) {

